### PR TITLE
Add authentication/authorization tests

### PR DESCRIPTION
**Pull request by cailyncodes**
Add tests of the authentication and authorization system, both unit and integration.

### CODE REVIEW

**Comment by cailyncodes** [comment]
These tests are passing locally. @speedcm what environment do we run the tests in? "Development" or something else?

**Comment by jlhogan** [comment]
> These tests are passing locally. @speedcm what environment do we run the tests in? "Development" or something else?

Tests should be running in "development" environment -- that's the default for dotnet unless something else is specified; they're run by the docker-compose project: "docker-compose.test.yaml"; have you tried running that locally to isolate?

File: src/Hedwig/Security/AuthorizationEvaluator.cs [hunk 1]
```diff
@@ -93,7 +93,7 @@ public AuthorizationEvaluator(AuthorizationSettings settings)
 					else
 					{
 						// Final rule must succeed or fail
-						throw new Exception();
+						throw new Exception("Final rule must by Allow or Deny");
```
Comment (jlhogan) [inline src/Hedwig/Security/AuthorizationEvaluator.cs]: nit: must *be*

File: test/HedwigTests/Security/AuthorizationRuleTests.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Http;
+using Moq;
+using Hedwig.Security;
+
+namespace HedwigTests.Security
+{
+    public class AuthorizationRuleTests
+    {
+        [Fact]
```
Comment (jlhogan) [inline test/HedwigTests/Security/AuthorizationRuleTests.cs]: you can write Xunit tests as "theories" which are just unit tests for which you can configure a data provider, for instance with an iterable set of auth rules, and expected results. Not sure if you tried that out !

File: test/HedwigTests/Security/Policies/CurrentUserRequirementTests.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        public async void When_Current_User_Authorize_Produces_No_Error()
+        {
+          // If
+          var requirement = new CurrentUserRequirement();
+          var context = new AuthorizationContext();
+          var claims = new Dictionary<string, string>() {
+            { "sub", "1" }
+          };
+          context.User = AuthorizationRequirementHelper.CreatePrincipal("password", claims);
+          var nameNode = new NameNode("id");
```
Comment (jlhogan) [inline test/HedwigTests/Security/Policies/CurrentUserRequirementTests.cs]: if not going to abstract this into some helper( totally fair!), might be useful to throw a comment explaining that this is simulating a graphQL request with query argument {id: 1}

File: test/HedwigTests/Integration/GraphQLQueries/ChildQueryPermissionsTests.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+				// Then
+				response.EnsureSuccessStatusCode();
+				GraphQLResponse res = await response.ParseGraphQLResponse();
+				Assert.NotNull(res.Errors);
+			}
+		}
+
+		[Fact]
+		public async Task When_Not_Developer_Get_Child_By_Id_Fails()
```
Comment (jlhogan) [inline test/HedwigTests/Integration/GraphQLQueries/ChildQueryPermissionsTests.cs]: does this mean right now the requirement for retrieve child data is developer role?

File: test/HedwigTests/Security/AuthorizationRuleTests.cs (hunk 2 above)
Comment (cailyncodes) [reply test/HedwigTests/Security/AuthorizationRuleTests.cs]: Oh did not know that. I'll have to check it out!

File: test/HedwigTests/Integration/GraphQLQueries/ChildQueryPermissionsTests.cs (hunk 4 above)
Comment (cailyncodes) [reply test/HedwigTests/Integration/GraphQLQueries/ChildQueryPermissionsTests.cs]: The team needs to have a discussion on roles and permissioning generally and so I just have it set as developer required for now, but should be changed once we hash out the requirements.

**Comment by cailyncodes** [comment]
> Tests should be running in "development" environment -- that's the default for dotnet unless something else is specified; they're run by the docker-compose project: "docker-compose.test.yaml"; have you tried running that locally to isolate?

Thanks for the tip! I needed to explicitly set the environment to development in `docker-compose.test.yaml`. The way I stubbed out an extension method for testing which environment it is required an non-implicit string.

File: test/HedwigTests/Security/AuthorizationRuleTests.cs (hunk 2 above)
Comment (cailyncodes) [reply test/HedwigTests/Security/AuthorizationRuleTests.cs]: Okay this is super cool!

File: test/HedwigTests/Security/Policies/CurrentUserRequirementTests.cs (hunk 3 above)
Comment (cailyncodes) [reply test/HedwigTests/Security/Policies/CurrentUserRequirementTests.cs]: Refactored!

**Pull request merged**